% Define the top matter
\setModuleTitle{Read Alignment}
\setModuleAuthors{%
  Sonika Tyagi\mailto{sonika.tyagi@agrf.org.au}
  Gayle Philip \mailto{unimelb.edu.au}
}
\setModuleContributions{%
  Mathieu Bourgey \mailto{mathieu.bourgey@mcgill.ca}%
}

%  Start: Module Title Page
\chapter{\moduleTitle}
\newpage
% End: Module Title Page

\section{Key Learning Outcomes}

After completing this practical the trainee should be able to:
\begin{itemize}
  \item Perform the simple NGS data alignment task against one interested reference data
  \item Learn about the SAM/BAM formats for further manipulation 
  \item Be able to sort and index BAM format for visualisation purposes.
\end{itemize}

\section{Resources You'll be Using}
 
\subsection{Tools Used}
\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=0.5cm]
%  \item[BWA Burrow wheel Algorithm]\hfill\\
  	\url{http://bio-bwa.sourceforge.net}
  \item[Samtools]\hfill\\
  	\url{http://picard.sourceforge.net/}
\end{description}

\section{Useful Links}
 
\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=0.5cm]
  \item[SAM Specification]\hfill\\
    \url{http://samtools.sourceforge.net/SAM1.pdf}
  \item[Explain SAM Flags]\hfill\\
    \url{http://picard.sourceforge.net/explain-flags.html}
\end{description}

\subsection{Sources of Data}
% TODO more specific about how the data was derived for this module
  \url{http://sra.dnanexus.com/studies/ERP001071}

\clearpage

\section{Introduction}

\begin{information}
The goal of this hands-on session is to perform an NGS alignment on the sequencing data coming from a tomour and normal group of samples. We will align raw sequencing data to the mouse genome using BWA aligner and then we will discuss at the the sequence alignment and mapping format (SAM). SAM to BAM coversion, indexing and sorting will also be demostrated. These are important and essential steps for downstreaming processing of the aligned BAM files. 
 
This data is the whole genome sequencing of a lung adenocarcinoma patient AK55. It was downloaded from ERP001071 (a link is provided in the data source section). Only the HiSeq2000 data for Blood and liverMets were analysed.

Accession numbers associated with read data assigned by the European Bioinformatics Institute (EBI) start with 'ER'. e.g. ERP is the study and ERR is the run. The original FASTQ files downloaded had the ERR number in front of each read name in the FASTQ file. The read name had to be edited to remove the ERR number at the start of the name. This had caused problems for downstream programs such as Picard for marking optical duplicates.

\end{information}

\section{Prepare the Environment}

\begin{information}
We will use one data set in this practical, which can be found in the \texttt{alignment}
directory on your desktop.
There were n \texttt{.fastq.gz} files from Blood and Liver groups that were used to perform the whole genome alignment using the BWA aligner. By now you know how a fastq file looks like. Next we will see how fastq files are aligned with the reference genome and what are the resulting standard alignment format. The actual alignment of all n files with human genome reference took x hrs and y minutes. In the interest of time we have selected only one file each from the each Blood and Liver group to run a BWA command. The remainining alignments have already been performed for you and will be required in the subsequent modules of the workshop.
\end{information}

\begin{steps}
Open the Terminal.

First, go to the right folder, where the data are stored.
\begin{lstlisting}
cd ~/alignment
\end{lstlisting}

\end{steps}

\section{Alignment}

\begin{information}
You already know that there are a number of competing tools for short read alignment, each with its own set of strengths, weaknesses, and caveats. Here we will use BWA, a widely used aligner based on Burrow Wheel Algorithm.  The alignment involves two steps- first, indexing the genome and then running the alignment command.  
\end{information}

\begin{steps}
BWA has a number of parameters in order to perform the alignment. To view them all type

\begin{lstlisting}
bwa <press enter> 
\end{lstlisting}

BWA uses indexed genome for the alignment in order to keep its memory footprint small. Because of time constraints we will not be running the indexing command. It is run only once for a version of genome, the complete command to index the human genome version hg19 is given below. 
FASTA format. This is stored in a file named \texttt{mm10}, under the
subdirectory \texttt{bowtie\_index}.

\begin{warning}
  You DO NOT need to run this command. This has already been run for you.
  \begin{lstlisting}
  bwa-build bwa_index/human_g1k_v37.fasta bwa_index/human_v37
  \end{lstlisting}
\end{warning}

This command will output 6 files that constitute the index. These files that have the prefix \texttt{human_v37} are stored in the \texttt{bwa\_index} subdirectory. To view if they files have been successfully created type:

\begin{lstlisting}
ls -l bwa_index
\end{lstlisting}
\end{steps}

\begin{information}
Now that the genome is indexed we can move on to the actual alignment. The first
argument for \texttt{bwa} is the basename of the index for the genome to be searched;
in our case this is \texttt{human_v37}. We also want to make sure that the output is
in SAM format using the \texttt{-S} parameter. The last argument is the name of the
FASTQ file.
\end{information}

\begin{steps}
Align the Oct4 reads using Bowtie2: 

\begin{lstlisting}
bwa mem -M -t 4 -R @RG\tSM:Blood\tID:ERR059354\tLB:lb\tPL:ILLUMINA human_g1k_v37.fasta Blood_ID_ERR059354_R1.fastq.gz Blood_ID_ERR059354_R2.fastq.gz > Blood_ID_ERR059354.sam
\end{lstlisting}

The above command outputs the alignment in SAM format and stores them in the
file \texttt{Blood_ID_ERR059354.sam}.
\end{steps}

\begin{note}
In general before you run Bowtie2, you have to know what quality encoding your FASTQ files
are in. The available FASTQ encodings for bowtie are:

\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
  \item[mem] =
  \item[-M] = flags extra hits as secondary. This is needed for compatibility with other tools downstream.
  \item[-t] = Number of threads.
  \item[-R] = Complete read group header line.
\end{description}

The FASTQ files we are working with are Sanger encoded (Phred+33), which is the
default for Bowtie2.

Bwa will take x-y minutes to align the file. This is fast compared to
other aligners which sacrifice some speed to obtain higher sensitivity.
\end{note}

\begin{steps}
Look at the top 10 lines of the SAM file by typing:

\begin{lstlisting}
head -n 10 Blood*.sam
\end{lstlisting}
\end{steps}

\begin{questions}
Can you distinguish between the header of the SAM format and the actual alignments?
\begin{answer}
The header line starts with the letter `@', i.e.: 

\begin{tabular}{lllll}
@HD & VN:1.0 & SO:unsorted & & \\
@SQ & SN:chr1 & LN:195471971 & & \\
@PG & ID:Bowtie2 &  PN:bowtie2   & VN:2.2.4  & CL:``/tools/bowtie2/bowtie2-default/bowtie2-align-s --wrapper basic-0 -x bowtie\_index/mm10 -q Oct4.fastq'' \\
\end{tabular}

While, the actual alignments start with read id, i.e.:

\begin{tabular}{llll}
SRR002012.45 & 0 & etc & \\
SRR002012.48 & 16 & chr1 & etc \\
\end{tabular}
\end{answer}

What kind of information does the header provide?
\begin{answer}
\begin{itemize}
  \item @HD: Header line; VN: Format version; SO: the sort order of alignments.
  \item @SQ: Reference sequence information; SN: reference sequence name; LN: reference sequence length.
  \item @PG: Read group information; ID: Read group identifier; VN: Program version; CL: the command line that produces the alignment.
\end{itemize}
\end{answer}

To which chromosome are the reads mapped? 
\begin{answer}
Chromosome 1.
\end{answer}
\end{questions}

\section{Manipulate SAM output}

\begin{note}
SAM files are rather big and when dealing with a high volume of NGS data,
storage space can become an issue. As we have already seen, we can convert SAM
to BAM files (their binary equivalent that are not human readable) that occupy
much less space.
\end{note}

\begin{steps}
Convert SAM to BAM using \texttt{samtools view} and store the output in the file
\texttt{Oct4.bam}. You have to instruct \texttt{samtools view} that the input is in SAM
format (\texttt{-S}), the output should be in BAM format (\texttt{-b}) and that
you want the output to be stored in the file specified by the \texttt{-o}
option:

\begin{lstlisting}
samtools view -bSo Oct4.bam Oct4.sam
\end{lstlisting}
\end{steps}

\begin{advanced}
Compute summary stats for the Flag values associated with the alignments using:

\begin{lstlisting}
samtools flagstat Oct4.bam
\end{lstlisting}
\end{advanced}

\section{Post alignment processing}

\begin{information}
IGV is a stand-alone genome browser. Please check their website
(\url{http://www.broadinstitute.org/igv/}) for all the formats that IGV
can display. For our visualization purposes we will use the BAM and bigWig
formats.
\end{information}

\begin{note}
When uploading a BAM file into the genome browser, the browser will look for the
index of the BAM file in the same folder where the BAM files is. The index file
should have the same name as the BAM file and the suffix \texttt{.bai}. Finally, to
create the index of a BAM file you need to make sure that the file is sorted
according to chromosomal coordinates.
\end{note}

\begin{steps}
Sort alignments according to chromosomal position and store the result in the
file with the prefix \texttt{Oct4.sorted}:

\begin{lstlisting}
samtools sort Oct4.bam Oct4.sorted
\end{lstlisting}

Index the sorted file.

\begin{lstlisting}
samtools index Oct4.sorted.bam
\end{lstlisting}

The indexing will create a file called \texttt{Oct4.sorted.bam.bai}. Note that
you don't have to specify the name of the index file when running
\texttt{samtools index}, it simply appends a \texttt{.bai} suffix to the input
BAM file.
\end{steps}

\begin{note}
\end{note}

\begin{steps}

\begin{lstlisting}
\end{lstlisting}
\end{steps}

\begin{note}
\end{note}

\begin{steps}

\begin{lstlisting}[style=command_syntax]
\end{lstlisting}


\begin{questions}
What is the 
\begin{answer}
The 
\end{answer}
\end{questions}


\begin{questions}
What do you think the different colors mean?
\begin{answer}
The different color represents four nucleotides, e.g. blue is Cytidine (C), red
is Thymidine (T).
\end{answer}
\end{questions}

