Key Learning Outcomes
---------------------

After completing this practical the trainee should be able to:

-   Perform the simple NGS data alignment task against reference data.

-   Learn about the SAM/BAM formats for further manipulation.

-   Be able to sort and index BAM format for visualisation purposes.

Resources You’ll be Using
-------------------------

### Tools Used

BWA Burrows-Wheeler Algorithm
:   \
    <http://bio-bwa.sourceforge.net>

Samtools
:   \
    <http://picard.sourceforge.net/>

Useful Links
------------

SAM Specification
:   \
    <http://samtools.sourceforge.net/SAM1.pdf>

Explain SAM Flags
:   \
    <http://picard.sourceforge.net/explain-flags.html>

### Sources of Data

<http://sra.dnanexus.com/studies/ERP001071>

Introduction
------------

The goal of this hands-on session is to perform an NGS alignment on the
sequencing data coming from a tumour and normal group of samples. We
will align raw sequencing data to the human genome using the BWA aligner
and then we will discuss the sequence alignment and mapping format
(SAM). SAM to BAM coversion, indexing and sorting will also be
demonstrated. These are important and essential steps for downstream
processing of the aligned BAM files.

This data is the whole genome sequencing of a lung adenocarcinoma
patient AK55. It was downloaded from `ERP001071` (a link is provided in
the data source section). Only the HiSeq2000 data for `Blood` and
`liverMets` were analysed.

Accession numbers associated with read data assigned by the European
Bioinformatics Institute (EBI) start with ’ER’. e.g. ERP is the study ID
and ERR is the run. The original FASTQ files downloaded had the ERR
number in front of each read name in the FASTQ file. The read name had
to be edited to remove the ERR number at the start of the name. This had
caused problems for downstream programs such as Picard for marking
optical duplicates.

We have used 4 Blood samples (8 PE `.fastq.gz` files) and 5 Liver
samples (10 PE `.fastq.gz` files) data from this study to perform the
whole genome alignment using the BWA aligner. The whole process took
\>150K CPU seconds per sample and the precomputed alignment will be used
in different sections of this workshop.

Prepare the Environment
-----------------------

By now you know about the raw sequence fastq format generated by the
Illumina sequencers. Next we will see how fastq files are aligned with
the reference genome and what are the resulting standard alignment
format. In the interest of time we have selected only 1 million paired
reads from a Blood sample to demonstrate a BWA command. The remainining
alignments have already been performed for you and will be required in
the subsequent modules of the workshop.

The input data for this section can be found in the `alignment`
directory on your desktop. Please follow the commands below to go to the
right folder and view top 10 lines of the input fastq file:

Open the Terminal.

First, go to the right folder, where the data are stored.

    cd /home/trainee/alignment
    ls
    zless input/SM_Blood_ID_ERR059356.subset_R1.fastq.gz

Alignment
---------

You already know that there are a number of competing tools for short
read alignment, each with its own set of strengths, weaknesses, and
caveats. Here we will use BWA, a widely used aligner based on the
Burrows-Wheeler Algorithm. The alignment involves two steps- first,
indexing the genome and then running the alignment command. BWA is a
software package for mapping low-divergent sequences against a large
reference genome, such as the human genome. It consists of three
algorithms: BWA-backtrack, BWA-SW and BWA-MEM. The first algorithm is
designed for Illumina sequence reads up to 100bp, while the rest two for
longer sequences ranged from 70bp to 1Mbp. BWA-MEM and BWA-SW share
similar features such as long-read support and split alignment, but
BWA-MEM, which is the latest, is generally recommended for high-quality
queries as it is faster and more accurate. BWA-MEM also has better
performance than BWA-backtrack for 70-100bp Illumina reads (ref:bwa
manual).

BWA has a number of parameters in order to perform the alignment. To
view them all type

    bwa <press enter> 

BWA uses an indexed genome for the alignment in order to keep its memory
footprint small. Because of time constraints we will not be running the
indexing command. It is run only once for a version of genome, and the
complete command to index the human genome version hg19 is given below.

You DO NOT need to run this command. This has already been run for you.

      bwa index -p bwaIndex/human_g1k_v37.fasta -a bwtsw human_g1k_v37.fasta
      

This command will output 6 files that constitute the index. These files
that have the prefix `human_v37` are stored in the `bwa_index`
subdirectory. To view if they files have been successfully created type:

We have used the following arguments for the indexing of the genome.

-p
:   = Prefix of the output database [same as db filename]

-a
:   = Algorithm for constructing BWT index. This method works with the
    whole human genome.

ref genome filename
:   = the last argument is the name of the reference genome file in the
    fasta format

<!-- -->

    ls -l bwaIndex

Now that the genome is indexed we can move on to the actual alignment.
The first argument for `bwa` is the basename of the index for the genome
to be searched; in our case this is `human_g1k_v37.fasta`.

Align the reads from Blood samples using the following command:

    bwa mem -M -t 4 -R @RG\tSM:Blood\tID:ERR059354\tLB:lb\tPL:ILLUMINA human_g1k_v37 SM_Blood_ID_ERR059356.subset_R1.fastq.gz SM_Blood_ID_ERR059356.subset_R2.fastq.gz > SM_Blood.sam

The above command outputs the alignment in SAM format and stores them in
the file `SM_Blood.sam`.

We have used the following arguments for the alignment of the reads.

mem
:   = fast mode of high quality input such the Illumina

-M
:   = flags extra hits as secondary. This is needed for compatibility
    with other tools downstream.

-t
:   = Number of threads.

-R
:   = Complete read group header line.

The SAM (Sequence Alignment/Map) format is currently the de facto
standard for storing large nucleotide sequence alignments. It is a
TAB-delimited text format consisting of a header section, which is
optional, and an alignment section. If present, the header must be prior
to the alignments. Header lines start with `@`, while alignment lines do
not. Each alignment line has 11 mandatory fields for essential alignment
information such as mapping position.

Look at the top 10 lines of the SAM file by typing:

    head -n 10 SM_Blood.sam

Can you distinguish between the header of the SAM format and the actual
alignments?

The header line starts with the letter ‘@’, i.e.:

  ----- ------------ -------------- ---------- ----------------------------------------------------------------------------------------------------------
  @HD   VN:1.0       SO:unsorted               
  @SQ   SN:chr1      LN:195471971              
  @PG   ID:Bowtie2   PN:bowtie2     VN:2.2.4   CL:“/tools/bowtie2/bowtie2-default/bowtie2-align-s –wrapper basic-0 -x bowtie\_index/mm10 -q Oct4.fastq”
  ----- ------------ -------------- ---------- ----------------------------------------------------------------------------------------------------------

While, the actual alignments start with read id, i.e.:

  -------------- ---- ------ -----
  SRR002012.45   0    etc    
  SRR002012.48   16   chr1   etc
  -------------- ---- ------ -----

What kind of information does the header provide?

-   @HD: Header line; VN: Format version; SO: the sort order of
    alignments.

-   @SQ: Reference sequence information; SN: reference sequence name;
    LN: reference sequence length.

-   @PG: Read group information; ID: Read group identifier; VN: Program
    version; CL: the command line that produces the alignment.

To which chromosome are the reads mapped?

Chromosome 1.

Manipulate SAM output
---------------------

SAM files are rather big and when dealing with a high volume of NGS
data, storage space can become an issue. As we have already seen, we can
convert SAM to BAM files (their binary equivalent that are not human
readable) that occupy much less space.

Convert SAM to BAM using `samtools view` and store the output in the
file `SM_Blood.bam`. You have to instruct `samtools view` that the input
is in SAM format (`-S`), the output should be in BAM format (`-b`) and
that you want the output to be stored in the file specified by the `-o`
option:

    samtools view -bSo SM_Blood.bam SM_Blood.sam

Compute summary stats for the Flag values associated with the alignments
using:

    samtools flagstat SM_Blood.bam

Post Alignment Visualisation option
-----------------------------------

IGV is a stand-alone genome browser that can be used to visualise the
BAM outputs.. Please check their website
(<http://www.broadinstitute.org/igv/>) for all the formats that IGV can
display.

We will be using IGV later in the workshop for viewing a BAM file in the
genome browser. It requires the index of the BAM file to be in the same
folder as where the BAM file is. The index file should have the same
name as the BAM file and the suffix `.bai`. Finally, to create the index
of a BAM file you need to make sure that the file is sorted according to
chromosomal coordinates.

Sort alignments according to chromosomal position and store the result
in the file with the prefix `Blood.sorted`:

    samtools sort SM_Blood.bam SM_Blood.sorted

Index the sorted file.

    samtools index SM_Blood.sorted.bam

The indexing will create a file called `Blood.sorted.bam.bai`. Note that
you don’t have to specify the name of the index file when running
`samtools index`, it simply appends a `.bai` suffix to the input BAM
file.

How can you quickly find out whether a BAM file is already coordinate
sorted or not ?

Use samtools view command to look at the SAM header.
